src/include/function/gds/gds_frontier.h

```diff
@@ -79,66 +161,57 @@ class GDSFrontier {
         numApproxActiveNodesForNextIter.fetch_add(i);
     }
     void beginNewIteration() {
         std::unique_lock<std::mutex> lck{mtx};
-        // If curIter is INVALID_IDX (which should be UINT32_MAX), which indicates that the
-        // iterations have not started, the following line will set it to 0.
         curIter.fetch_add(1u);
+        numApproxActiveNodesForCurLevel.store(numApproxActiveNodesForNextIter.load());
         numApproxActiveNodesForNextIter.store(0u);
+        std::swap(curFrontier, nextFrontier);
         beginNewIterationInternalNoLock();
     }
-    // When performing computations on multi-label graphs, it may be beneficial to fix a single
-    // node table of nodes in the current frontier and a single node table of nodes for the next
-    // frontier. That is because algorithms will generally perform extensions using a single
-    // relationship table at a time, and each relationship table R is between a single source node
-    // table S and a single destination node table T. Therefore, generally during execution the
-    // algorithm will need to check only the active S nodes in current frontier and update the
-    // active statuses of only the T nodes in the next frontier. The information that the algorithm
-    // is beginning and S-to-T extensions can be given to the frontiers to possibly avoid them
-    // doing lookups of S and T-related data structures, e.g., masks, internally.
-    // Note that ideally a similar function, say called "fixNodeTable", could be in GDSFrontier,
-    // since Frontiers has two GDSFrontier instances in it. However, some implementations of
-    // Frontiers can use a single GDSFrontier to mimic two frontiers (e.g., identifying nodes
-    // in current or next frontier by specific mask values. For such implementations we would have
-    // to call fixNodeTable(S) and fixNodeTable(T) on the same object, which could overwrite
-    // each other. That is why this function is put in the Frontiers class.
+    virtual void initRJFromSource(nodeID_t source) = 0;
     virtual void beginFrontierComputeBetweenTables(table_id_t curFrontierTableID,
         table_id_t nextFrontierTableID) = 0;
-    idx_t getNextIter() {
-        // Note: If curIter is INVALID_IDX (which should be UINT32_MAX), which indicates that the
-        // iterations have not started, this will return 0.
-        return curIter.load() + 1u;
-    }
-    bool hasActiveNodesForNextIter() { return numApproxActiveNodesForNextIter.load() > 0; }
+
+    idx_t getNextIter() { return curIter.load() + 1u; }
+
+    bool hasActiveNodesForNextLevel() { return numApproxActiveNodesForNextIter.load() > 0; }
+
     // Note: If the implementing class stores 2 frontiers, this function should swap them.
     virtual void beginNewIterationInternalNoLock() {}
 
 protected:
     std::mutex mtx;
-    std::atomic<idx_t> curIter;
+    // curIter is the iteration number of the algorithm and starts from 0.
+    std::atomic<uint16_t> curIter;
```

Comment:
Yes, this is read in different classes and by different threads. I need the lock in the constructor because I need multiple fields to be updated atomically.